app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -814,6 +824,26 @@ public void onRequestPermissionsResult(int requestCode,
                     scanSMSLinks(manualSMSInstall);
                 }
             }
+        } else if (requestCode == Permissions.ALL_PERMISSIONS_REQUEST) {
+            String[] requiredPerms = Permissions.getRequiredPerms();
+
+            for (int i = 0; i < permissions.length; i++) {
+                for (String requiredPerm : requiredPerms) {
+                    if (requiredPerm.equals(permissions[i]) &&
+                            grantResults[i] == PackageManager.PERMISSION_DENIED) {
+                        showMissingPermissionState();
+                        return;
+                    }
+                }
+            }
         }
     }
+
+    private void showMissingPermissionState() {
+        // TODO PLM: instead of popping up the same message we should disable
+        // install buttons and show a message and button to re-request the
+        // permissions.
+        Permissions.acquireAllAppPermissions(this, this,

```

Comment:
The app won't work if it doesn't have these permissions, so for now, the lowest bar for this state is just re-ask for perms. The TODO outlines a future state that is more user friendly.